File: base/src/main/java/com/boilerplate/base/mvp/BasePresenter.java
package com.boilerplate.base.mvp;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;


public class BasePresenter<T extends IView> implements IPresenter<T> {


    protected T view;
    protected CompositeDisposable mCompositeDisposable;

    @Override
    public void attachView(T view) {
        this.view = view;
    }

    protected void unSubscribe() {
        if (mCompositeDisposable != null) {
            mCompositeDisposable.clear();
            mCompositeDisposable.dispose();
        }
    }

    protected void addSubscribe(Disposable disposable) {
        if (mCompositeDisposable == null) {
            mCompositeDisposable = new CompositeDisposable();
        }
        mCompositeDisposable.add(disposable);
    }

    @Override
    public void detachView() {
        this.view = null;
        unSubscribe();
    }
}

File: base/src/main/java/com/boilerplate/base/kotlin/BasePresenter.kt
package com.superc.base

import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable

/**
 * Created by Louis on 2018/4/27.
 */
class BasePresenter<T : IView> : IPresenter<T> {

    override fun attachView(view: T) {
        this.view = view
    }


    protected var mCompositeDisposable: CompositeDisposable? = null


    override fun detachView() {
    }

    protected fun unSubsrcibe() {
        if (mCompositeDisposable != null) {
            mCompositeDisposable!!.clear()
            mCompositeDisposable!!.dispose()
        }
    }

    protected fun addSubscribe(disposable: Disposable) {
        if (mCompositeDisposable == null) {
            mCompositeDisposable = CompositeDisposable()
        }
        mCompositeDisposable!!.add(disposable)
    }

    protected var view: T? = null
}